File: 백준/Silver/14940. 쉬운 최단거리/쉬운 최단거리.py
import sys
from collections import deque
input = sys.stdin.readline

def bfs():
    que = deque()

    for i in range(n):
        for j in range(m):
            if graph[i][j] == 2:
                que.append((i, j))
                graph[i][j] = 0

                while que:
                    x, y = que.popleft()
                    visited[x][y] = True

                    for i in range(4):
                        nx = x + dxy[i][0]
                        ny = y + dxy[i][1]

                        if nx < 0 or nx >= n or ny < 0 or ny >= m:
                            continue
                        if not visited[nx][ny] and graph[nx][ny] == 1:
                            visited[nx][ny] = True
                            graph[nx][ny] = graph[x][y] + 1
                            que.append((nx, ny))

                return


n, m = map(int, input().split())
graph = []
ans = [[-1] * m for _ in range(n)]
dxy = [(1, 0), (0, 1), (-1, 0), (0, -1)]
visited = [[False] * m for _ in range(n)]

for i in range(n):
    graph.append(list(map(int, input().split())))

bfs()

for i in range(n):
    for j in range(m):
        if graph[i][j] == 1 and not visited[i][j]:
            print(-1, end=' ')
        else:
            print(graph[i][j], end=' ')
    print()

File: 백준/Silver/14940. 쉬운 최단거리/쉬운 최단거리.java
import java.io.*;
import java.util.*;


public class Main {
	static BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
	static StringBuilder output = new StringBuilder();
	static StringTokenizer tokens;
	static int N, M;
	static int[][] map, visited;
	static int[][] deltas = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
	
	public static void main(String[] args) throws IOException {
		tokens = new StringTokenizer(input.readLine());

		N = Integer.parseInt(tokens.nextToken());
		M = Integer.parseInt(tokens.nextToken());
		map = new int[N][M];
		visited = new int[N][M];
		int r = 0;
		int c = 0;
		
		for(int i = 0; i < N; i++) {
			for(int j = 0; j < M; j++) {
				visited[i][j] = -1; 
			}
		}
		
		for(int i = 0; i < N; i++) {
			tokens = new StringTokenizer(input.readLine());
			for(int j = 0; j < M; j++) {
				map[i][j] = Integer.parseInt(tokens.nextToken());
				
				if(map[i][j]== 0) visited[i][j]= 0; 
				if(map[i][j]== 2 ) {
					r = i;
					c = j;
				}
			}
		}
		
		bfs(r, c);
		
		for(int i = 0; i < N; i++) {
			for(int j = 0; j < M; j++) {
				output.append(visited[i][j] + " " );
			}
			output.append("\n");
		}
		
		System.out.println(output);
	}
	
	static void bfs(int r, int c) {
		Deque<Pos> que = new ArrayDeque<>();
		visited[r][c] = 0;
		que.add(new Pos(r, c));
		
		while(!que.isEmpty()) {
			Pos now = que.poll();
			
			for(int d = 0; d < 4; d++) {
				int nr = now.r + deltas[d][0];
				int nc = now.c + deltas[d][1];
								
				if(nr < 0 || nc < 0 || nr >= N || nc >= M) continue;
				
				if(visited[nr][nc] == -1) {
					visited[nr][nc]= visited[now.r][now.c] + 1;
					que.add(new Pos(nr, nc));
				}
				
				
			}
		}
	}
	
	static class Pos {
		int r;
		int c;
		
		public Pos(int r, int c) {
			super();
			this.r = r;
			this.c = c;
		}
	}
}